### PR TITLE
Propagate permission changes to federated servers

### LINKED ISSUE #12953: Federated users are not aware of changed permissions
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

---

## Steps to reproduce
1. Invite a federated user and accept the invite as a federated user
2. Change the permissions of the user to not allow voice and video
3. Check the permission on the federated side, the room object still has all permissions set

Only after starting a call, we get the correct permissions via signaling message and then reconnect immediately.

## Talk app

**Talk app version:** 20 Beta 3

### PR DESCRIPTION
**Pull request by danxuliu**
Fix #12953
Requires (for the integration tests to pass) #13027

The frontend gets the permission for the current participant from the room data (as guest users can not get it from the participant list). However, permission changes were not propagated to federated servers, so the frontend of federated users always assumed that the user had the default participant permissions.

However, note that media related permissions (audio, video and screenshare) were correctly handled (at least in the WebUI), as when the permissions are modified a `participants->update` signaling message is sent and the WebUI gets the media permissions directly from that signaling message rather than from the room data.

Independently of that, note that it is not enough to just propagate participant permissions, room permissions and even general default permissions need to be propagated as well (but default permissions currently are not, see below), as when there are no custom permissions for a participant [it falls back to the room permissions](https://github.com/nextcloud/spreed/blob/1ccbf67da231c37801c1406b10000144125aec35/lib/Participant.php#L119-L136).

## TODO
- [X] Fix propagating permissions when changed in the conversation
- [ ] Currently this pull request propagates participant and room permissions, but not the default permissions in the host server. Those are [just read from a configuration value](https://github.com/nextcloud/spreed/blob/1ccbf67da231c37801c1406b10000144125aec35/lib/Config.php#L258), and as far as I know there is no event to listen to in case it changes. I do not know how to propagate those, maybe through capabilities?
- [ ] For simplicity the resource type `TALK_ROOM_RESOURCE` and, therefore, the `CloudFederationProviderTalk` were reused for participant related notifications. Should a new `TALK_PARTICIPANT_RESOURCE` and provider be added instead?

## How to test (scenario 1)

- Setup Nextcloud server A
- Setup Nextcloud server B
- In Nextcloud server A, create a conversation
- Add a user from Nextcloud server B
- Set custom permissions for the federated user and disable starting a call and sending chat messages
- In a private window, log in as the federated user
- Accept the invitation
- Open the federated conversation

### Result with this pull request

The new message area and the "Start call" button are disabled

### Result without this pull request

Neither the new message area nor the "Start call" button are disabled

## How to test (scenario 2)

- Setup Nextcloud server A
- Setup Nextcloud server B
- In Nextcloud server A, create a conversation
- Set custom permissions for the conversation and disable starting a call and sending chat messages
- Add a user from Nextcloud server B
- In a private window, log in as the federated user
- Accept the invitation
- Open the federated conversation

### Result with this pull request

The new message area and the "Start call" button are disabled

### Result without this pull request

Neither the new message area nor the "Start call" button are disabled

### CODE REVIEW

File: lib/Federation/BackendNotifier.php [hunk 1]
```diff
@@ -101,6 +103,8 @@ public function sendRemoteShare(
 		$protocol['invitedCloudId'] = $invitedCloudId->getId();
 		$protocol['roomName'] = $roomName;
 		$protocol['roomType'] = $roomType;
+		$protocol['roomCallPermissions'] = $roomCallPermissions;
+		$protocol['roomDefaultPermissions'] = $roomDefaultPermissions;
```
Comment (nickvergessen) [inline lib/Federation/BackendNotifier.php:107]: And description, and call state, and, … 

No this doesn't scale.
And as said, we remove call permissions, please don't add it

File: lib/Federation/CloudFederationProviderTalk.php [hunk 2]
```diff
@@ -125,6 +125,8 @@ public function shareReceived(ICloudFederationShare $share): string {
 		$remoteId = $share->getProviderId();
 		$roomToken = $share->getResourceName();
 		$roomName = $share->getProtocol()['roomName'];
+		$roomCallPermissions = $share->getProtocol()['roomCallPermissions'];
```
Comment (nickvergessen) [inline lib/Federation/CloudFederationProviderTalk.php:128]:
```suggestion
```

File: lib/Federation/CloudFederationProviderTalk.php [hunk 3]
```diff
@@ -322,6 +362,10 @@ private function roomModified(int $remoteAttendeeId, array $notification): array
 			}
 		} elseif ($notification['changedProperty'] === ARoomModifiedEvent::PROPERTY_AVATAR) {
 			$this->roomService->setAvatar($room, $notification['newValue']);
+		} elseif ($notification['changedProperty'] === ARoomModifiedEvent::PROPERTY_CALL_PERMISSIONS) {
+			$this->roomService->setPermissions($room, 'call', $notification['method'], $notification['permissions'], $notification['resetCustomPermissions']);
```
Comment (nickvergessen) [inline lib/Federation/CloudFederationProviderTalk.php:366]:
```suggestion
```

File: lib/Federation/FederationManager.php [hunk 4]
```diff
@@ -74,6 +77,8 @@ public function addRemoteRoom(
 		int $remoteAttendeeId,
 		int $roomType,
 		string $roomName,
+		int $roomCallPermissions,
```
Comment (nickvergessen) [inline lib/Federation/FederationManager.php:80]:
```suggestion
```

File: lib/Federation/FederationManager.php [hunk 5]
```diff
@@ -90,6 +95,14 @@ public function addRemoteRoom(
 			$room = $this->manager->createRemoteRoom($roomType, $roomName, $remoteToken, $remoteServerUrl);
 		}
 
+		// Only update the room permissions if there are no participants in the
+		// remote room. Otherwise the room permissions would be up to date
+		// already due to the notifications about room permission changes.
+		if (empty($participant = $this->participantService->getParticipantsForRoom($room))) {
+			$this->roomService->setPermissions($room, 'call', Attendee::PERMISSIONS_MODIFY_SET, $roomCallPermissions, true);
```
Comment (nickvergessen) [inline lib/Federation/FederationManager.php:102]:
```suggestion
```

File: lib/Federation/FederationManager.php [hunk 6]
```diff
@@ -90,6 +95,14 @@ public function addRemoteRoom(
 			$room = $this->manager->createRemoteRoom($roomType, $roomName, $remoteToken, $remoteServerUrl);
 		}
 
+		// Only update the room permissions if there are no participants in the
+		// remote room. Otherwise the room permissions would be up to date
+		// already due to the notifications about room permission changes.
+		if (empty($participant = $this->participantService->getParticipantsForRoom($room))) {
```
Comment (nickvergessen) [inline lib/Federation/FederationManager.php:101]: 1. No inline assignments
2. Use `getNumberOfActors` if you only need a count

File: lib/Federation/Proxy/TalkV1/Notifier/RoomModifiedListener.php [hunk 7]
```diff
@@ -42,13 +44,16 @@ public function handle(Event $event): void {
...
 				&& !$event instanceof LobbyModifiedEvent
+				&& !$event instanceof PermissionsModifiedEvent
 				&& !$event instanceof RoomModifiedEvent) {
 			return;
 		}
 
 		if (!in_array($event->getProperty(), [
 			ARoomModifiedEvent::PROPERTY_ACTIVE_SINCE,
 			ARoomModifiedEvent::PROPERTY_AVATAR,
+			ARoomModifiedEvent::PROPERTY_CALL_PERMISSIONS,
```
Comment (nickvergessen) [inline lib/Federation/Proxy/TalkV1/Notifier/RoomModifiedListener.php:55]:
```suggestion
```

File: tests/php/Federation/FederationTest.php [hunk 8]
```diff
@@ -256,6 +256,8 @@ public function testReceiveRemoteShare(): void {
 		$shareType = 'user';
 		$roomType = Room::TYPE_GROUP;
 		$roomName = 'Room name';
+		$roomCallPermissions = Attendee::PERMISSIONS_CUSTOM | Attendee::PERMISSIONS_PUBLISH_AUDIO;
```
Comment (nickvergessen) [inline tests/php/Federation/FederationTest.php:259]:
```suggestion
```

File: tests/php/Federation/FederationTest.php [hunk 9]
```diff
@@ -277,6 +279,8 @@ public function testReceiveRemoteShare(): void {
 			'name' => 'nctalk',
 			'roomType' => $roomType,
 			'roomName' => $roomName,
+			'roomCallPermissions' => $roomCallPermissions,
```
Comment (nickvergessen) [inline tests/php/Federation/FederationTest.php:282]:
```suggestion
```

File: tests/php/Federation/FederationTest.php [hunk 10]
```diff
@@ -288,7 +292,7 @@ public function testReceiveRemoteShare(): void {
 		// Test receiving federation expectations
 		$this->federationManager->expects($this->once())
 			->method('addRemoteRoom')
-			->with($shareWithUser, $providerId, $roomType, $roomName, $name, $remote, $token)
+			->with($shareWithUser, $providerId, $roomType, $roomName, $roomCallPermissions, $roomDefaultPermissions, $name, $remote, $token)
```
Comment (nickvergessen) [inline tests/php/Federation/FederationTest.php:295]:
```suggestion
			->with($shareWithUser, $providerId, $roomType, $roomName, $roomDefaultPermissions, $name, $remote, $token)
```

File: lib/Federation/Proxy/TalkV1/Notifier/ParticipantModifiedListener.php [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+			AParticipantModifiedEvent::PROPERTY_PERMISSIONS,
+		], true)) {
+			return;
+		}
+
+		$participants = $this->participantService->getParticipantsByActorType($event->getRoom(), Attendee::ACTOR_FEDERATED_USERS);
+		foreach ($participants as $participant) {
+			$cloudId = $this->cloudIdManager->resolveCloudId($participant->getAttendee()->getActorId());
+
+			$success = $this->notifyParticipantModified($cloudId, $participant, $event);
```
Comment (nickvergessen) [inline lib/Federation/Proxy/TalkV1/Notifier/ParticipantModifiedListener.php:49]: This should only be sent to the instance of the user that is modified.
We do not tell non-moderators the permissions of other users?!

File: lib/Federation/BackendNotifier.php [hunk 12]
```diff
@@ -69,6 +69,8 @@ public function sendRemoteShare(
 		$roomName = $room->getName();
 		$roomType = $room->getType();
 		$roomToken = $room->getToken();
+		$roomCallPermissions = $room->getCallPermissions();
```
Comment (nickvergessen) [inline lib/Federation/BackendNotifier.php:72]:
```suggestion
```

File: lib/Federation/BackendNotifier.php [hunk 13]
```diff
@@ -101,6 +103,8 @@ public function sendRemoteShare(
 		$protocol['invitedCloudId'] = $invitedCloudId->getId();
 		$protocol['roomName'] = $roomName;
 		$protocol['roomType'] = $roomType;
+		$protocol['roomCallPermissions'] = $roomCallPermissions;
```
Comment (nickvergessen) [inline lib/Federation/BackendNotifier.php:106]:
```suggestion
```

File: lib/Federation/CloudFederationProviderTalk.php [hunk 14]
```diff
@@ -294,7 +298,43 @@ private function shareUnshared(int $remoteAttendeeId, array $notification): arra
...
+		}
+
+		try {
+			$participant = $this->participantService->getParticipant($room, $invite->getUserId());
+		} catch (ParticipantNotFoundException $e) {
+			throw new ShareNotFound(FederationManager::OCM_RESOURCE_NOT_FOUND);
+		}
+
+		if ($notification['changedProperty'] === AParticipantModifiedEvent::PROPERTY_PERMISSIONS) {
+			$this->participantService->updatePermissions($room, $participant, Attendee::PERMISSIONS_MODIFY_SET, $notification['newValue']);
```
Comment (nickvergessen) [inline lib/Federation/CloudFederationProviderTalk.php:327]: this always updates the permissions of the recipient, independent of which users permissions were actually modified? :D

Comment (danxuliu) [reply lib/Federation/CloudFederationProviderTalk.php:327]: The event includes [the final permissions, no matter if they were set, added or removed](https://github.com/nextcloud/spreed/blob/d23239a03bfbe3086c5551bc86d619cb3d3ff512/lib/Service/ParticipantService.php#L206-L214), so the effect should be the same, no? :thinking:

File: lib/Federation/BackendNotifier.php (hunk 1 above)
Comment (danxuliu) [reply lib/Federation/BackendNotifier.php:107]: > And description, and call state, and, …
> 
> No this doesn't scale.

Well, then some other approach is needed. But [the initial state needs to be propagated somehow](https://github.com/nextcloud/spreed/issues/13060).

> And as said, we remove call permissions, please don't add it

This code was written before it was decided that call permissions were going to be removed. Anyway, I will redo it and rebase once call permissions are removed in _main_ and _stable30_.

**Comment by danxuliu** [comment]
Replaced by #13092

**Pull request closed without merging**